File: sdk-java-blocking/src/test/java/dev/restate/sdk/blocking/CompensationTest.java
package dev.restate.sdk.blocking;

import static dev.restate.sdk.core.impl.ProtoUtils.greetingRequest;
import static dev.restate.sdk.core.impl.ProtoUtils.greetingResponse;

import dev.restate.sdk.core.StateKey;
import dev.restate.sdk.core.TypeTag;
import dev.restate.sdk.core.impl.CompensationTestSuite;
import dev.restate.sdk.core.impl.testservices.GreeterGrpc;
import dev.restate.sdk.core.impl.testservices.GreetingRequest;
import dev.restate.sdk.core.impl.testservices.GreetingResponse;
import io.grpc.BindableService;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;

public class CompensationTest extends CompensationTestSuite {

  private static class ThrowManually extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {
    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      RestateContext ctx = restateContext();

      ctx.compensate(() -> ctx.set(StateKey.of("message", TypeTag.STRING_UTF8), "none"));

      throw Status.FAILED_PRECONDITION.asRuntimeException();
    }
  }

  @Override
  protected BindableService throwManually() {
    return new ThrowManually();
  }

  private static class NonTerminalErrorDoesntExecuteCompensations
      extends GreeterGrpc.GreeterImplBase implements RestateBlockingService {
    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      RestateContext ctx = restateContext();

      ctx.compensate(() -> ctx.set(StateKey.of("message", TypeTag.STRING_UTF8), "none"));

      throw new RuntimeException("non-terminal");
    }
  }

  @Override
  protected BindableService nonTerminalErrorDoesntExecuteCompensations() {
    return new NonTerminalErrorDoesntExecuteCompensations();
  }

  private static class CallCompensate extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {
    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      RestateContext ctx = restateContext();

      ctx.compensate(() -> ctx.set(StateKey.of("message", TypeTag.STRING_UTF8), "none"));
      GreetingResponse greetingResponse =
          ctx.call(GreeterGrpc.getGreetMethod(), greetingRequest("Francesco")).await();
      ctx.set(StateKey.of("message", TypeTag.STRING_UTF8), greetingResponse.getMessage());

      responseObserver.onNext(greetingResponse("ok"));
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService callCompensate() {
    return new CallCompensate();
  }

  private static class IllegalGetStateWithinCompensation extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {
    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      RestateContext ctx = restateContext();

      ctx.compensate(() -> ctx.get(StateKey.of("message", TypeTag.STRING_UTF8)));

      throw Status.FAILED_PRECONDITION.asRuntimeException();
    }
  }

  @Override
  protected BindableService illegalGetStateWithinCompensation() {
    return new IllegalGetStateWithinCompensation();
  }
}

File: sdk-kotlin/src/test/kotlin/dev/restate/sdk/kotlin/CompensationTest.kt
package dev.restate.sdk.kotlin

import dev.restate.sdk.core.StateKey
import dev.restate.sdk.core.TypeTag
import dev.restate.sdk.core.impl.CompensationTestSuite
import dev.restate.sdk.core.impl.ProtoUtils
import dev.restate.sdk.core.impl.testservices.*
import io.grpc.BindableService
import io.grpc.Status
import java.lang.UnsupportedOperationException
import kotlinx.coroutines.Dispatchers

class CompensationTest : CompensationTestSuite() {
  private class ThrowManually :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx = restateContext()
      ctx.compensate { ctx.set(StateKey.of("message", TypeTag.STRING_UTF8), "none") }
      throw Status.FAILED_PRECONDITION.asRuntimeException()
    }
  }

  override fun throwManually(): BindableService {
    return ThrowManually()
  }

  private class NonTerminalErrorDoesntExecuteCompensations :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx = restateContext()
      ctx.compensate { ctx.set(StateKey.of("message", TypeTag.STRING_UTF8), "none") }
      throw RuntimeException("non-terminal")
    }
  }

  override fun nonTerminalErrorDoesntExecuteCompensations(): BindableService {
    throw UnsupportedOperationException("https://github.com/restatedev/sdk-java/issues/116")
  }

  private class CallCompensate :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx = restateContext()
      ctx.compensate { ctx.set(StateKey.of("message", TypeTag.STRING_UTF8), "none") }
      val res = ctx.call(GreeterGrpcKt.greetMethod, ProtoUtils.greetingRequest("Francesco"))
      ctx.set(StateKey.of("message", TypeTag.STRING_UTF8), res.getMessage())
      return greetingResponse { message = "ok" }
    }
  }

  override fun callCompensate(): BindableService {
    return CallCompensate()
  }

  private class IllegalGetStateWithinCompensation :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx = restateContext()
      ctx.compensate { ctx.get(StateKey.of("message", TypeTag.STRING_UTF8)) }
      throw Status.FAILED_PRECONDITION.asRuntimeException()
    }
  }

  override fun illegalGetStateWithinCompensation(): BindableService {
    return IllegalGetStateWithinCompensation()
  }
}
